File: ts/test/regex-corpus.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import { describe, test } from 'node:test'
import Assert from 'node:assert'

import * as Fs from 'node:fs'
import * as Path from 'node:path'

import { normaliseRe } from '../dist/val/ConstraintVal'


const CORPUS = Path.join(
  __dirname, '..', '..', 'test', 'spec', 'files', 'regex-corpus.tsv')


type Row = { pattern: string, verdict: string, line: number }


function loadCorpus(): Row[] {
  const rows: Row[] = []
  const text = Fs.readFileSync(CORPUS, 'utf8')
    .replaceAll('\r\n', '\n').replaceAll('\r', '\n')
  let line = 0
  for (const raw of text.split('\n')) {
    line++
    if ('' === raw || raw.startsWith('#')) {
      continue
    }
    const tab = raw.indexOf('\t')
    rows.push({ pattern: raw.slice(0, tab), verdict: raw.slice(tab + 1), line })
  }
  return rows
}


describe('regex-corpus', () => {

  const rows = loadCorpus()

  test('corpus-is-loaded', () => {
    // A guard on the guard: an empty or truncated corpus would make
    // every assertion below vacuous while still reporting green.
    Assert.ok(300 < rows.length, 'corpus too small: ' + rows.length)
    Assert.ok(rows.some((r) => r.verdict.startsWith('!')),
      'corpus has no refusals — it would not exercise the subset bounds')
    Assert.ok(rows.some((r) => !r.verdict.startsWith('!')),
      'corpus has no acceptances')
  })


  test('verdict-parity', () => {
    for (const row of rows) {
      const [norm, why] = normaliseRe(row.pattern)
      const got = '' === why ? norm : '!' + why
      Assert.equal(got, row.verdict,
        'regex-corpus.tsv line ' + row.line + ': ' + JSON.stringify(row.pattern))
    }
  })


  test('accepted-patterns-compile', () => {
    for (const row of rows) {
      if (row.verdict.startsWith('!')) {
        continue
      }
      // The `u` flag is part of the contract, not the test: it is what
      // makes JavaScript match code points as RE2 does.
      Assert.doesNotThrow(() => new RegExp(row.verdict, 'u'),
        'normalised form does not compile, line ' + row.line + ': ' +
        JSON.stringify(row.verdict))
    }
  })


  test('normalisation-is-idempotent', () => {
    for (const row of rows) {
      if (row.verdict.startsWith('!')) {
        continue
      }
      const [again, why] = normaliseRe(row.verdict)
      Assert.equal(why, '',
        'normalised form is refused on re-normalisation, line ' + row.line +
        ': ' + JSON.stringify(row.verdict) + ' -> ' + why)
      Assert.equal(again, row.verdict,
        'normalisation is not idempotent, line ' + row.line)
    }
  })


  test('a-long-repeat-bound-is-refused', () => {
    for (const n of [1000, 100000]) {
      const [, why] = normaliseRe('x{' + '1'.repeat(n) + '}')
      Assert.match(why, /repeat count above/, n + ' digits: ' + why)
    }

    const [, comma] = normaliseRe('x{2,' + '9'.repeat(100000) + '}')
    Assert.match(comma, /repeat count above/)

    // A long run of LEADING ZEROS is a small number, not an over-cap
    // one.
    const zeros = 'x{' + '0'.repeat(100000) + '5}'
    const [out, why] = normaliseRe(zeros)
    Assert.equal(why, '')
    Assert.equal(out, zeros)
  })

})

File: go/regex_corpus_test.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu


import (
	"bufio"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
)

const regexCorpusPath = "../test/spec/files/regex-corpus.tsv"

type regexCorpusRow struct {
	pattern string
	verdict string
	line    int
}

func loadRegexCorpus(t *testing.T) []regexCorpusRow {
	t.Helper()
	f, err := os.Open(filepath.FromSlash(regexCorpusPath))
	if nil != err {
		t.Fatalf("cannot open %s: %v", regexCorpusPath, err)
	}
	defer f.Close()

	rows := []regexCorpusRow{}
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1024*1024), 1024*1024)
	line := 0
	for sc.Scan() {
		line++
		raw := sc.Text()
		if "" == raw || strings.HasPrefix(raw, "#") {
			continue
		}
		tab := strings.Index(raw, "\t")
		rows = append(rows, regexCorpusRow{
			pattern: raw[:tab], verdict: raw[tab+1:], line: line})
	}
	return rows
}

// TestRegexCorpusLoaded is a guard on the guard: an empty or truncated
// corpus would make every assertion below vacuous while reporting green.
func TestRegexCorpusLoaded(t *testing.T) {
	rows := loadRegexCorpus(t)
	if 300 >= len(rows) {
		t.Fatalf("corpus too small: %d", len(rows))
	}
	refused, accepted := false, false
	for _, r := range rows {
		if strings.HasPrefix(r.verdict, "!") {
			refused = true
		} else {
			accepted = true
		}
	}
	if !refused {
		t.Fatal("corpus has no refusals -- it would not exercise the subset bounds")
	}
	if !accepted {
		t.Fatal("corpus has no acceptances")
	}
}

// TestRegexCorpusVerdictParity is the cross-port check: the corpus is
// the pinned output of BOTH normalisers, so a drift in either fails on
// the exact line.
func TestRegexCorpusVerdictParity(t *testing.T) {
	for _, r := range loadRegexCorpus(t) {
		norm, why := normaliseRe(r.pattern)
		got := norm
		if "" != why {
			got = "!" + why
		}
		if got != r.verdict {
			t.Errorf("regex-corpus.tsv line %d: %q\n  got:  %q\n  want: %q",
				r.line, r.pattern, got, r.verdict)
		}
	}
}

// TestRegexCorpusAcceptedCompile checks that a normalisation never emits
// something this host refuses -- which would otherwise surface only as a
// runtime error much later.
func TestRegexCorpusAcceptedCompile(t *testing.T) {
	for _, r := range loadRegexCorpus(t) {
		if strings.HasPrefix(r.verdict, "!") {
			continue
		}
		if _, err := regexp.Compile(r.verdict); nil != err {
			t.Errorf("normalised form does not compile, line %d: %q: %v",
				r.line, r.verdict, err)
		}
	}
}

// TestRegexCorpusIdempotent catches an expansion that re-expands its own
// output, or that emits a construct the normaliser would itself refuse.
func TestRegexCorpusIdempotent(t *testing.T) {
	for _, r := range loadRegexCorpus(t) {
		if strings.HasPrefix(r.verdict, "!") {
			continue
		}
		again, why := normaliseRe(r.verdict)
		if "" != why {
			t.Errorf("normalised form refused on re-normalisation, line %d: %q -> %s",
				r.line, r.verdict, why)
			continue
		}
		if again != r.verdict {
			t.Errorf("normalisation is not idempotent, line %d: %q -> %q",
				r.line, r.verdict, again)
		}
	}
}

func TestLongRepeatBoundIsRefusedCheaply(t *testing.T) {
	for _, n := range []int{1000, 100000} {
		pattern := "x{" + strings.Repeat("1", n) + "}"
		out, why := normaliseRe(pattern)
		if "" == why {
			t.Fatalf("%d digits: accepted, normalised to %q", n, out)
		}
		if !strings.Contains(why, "repeat count above") {
			t.Fatalf("%d digits: %s", n, why)
		}
	}

	// And the same run with a comma in it, which takes the other exit.
	pattern := "x{2," + strings.Repeat("9", 100000) + "}"
	if _, why := normaliseRe(pattern); !strings.Contains(why, "repeat count above") {
		t.Fatalf("comma form: %s", why)
	}

	// A long run of LEADING ZEROS is a small number, not an over-cap
	// one: the fold saturates on value, exactly as `Atoi` did on the
	// assembled string.
	pattern = "x{" + strings.Repeat("0", 100000) + "5}"
	if out, why := normaliseRe(pattern); "" != why {
		t.Fatalf("leading zeros: %s", why)
	} else if out != pattern {
		t.Fatalf("leading zeros rewritten: %q", out)
	}
}
